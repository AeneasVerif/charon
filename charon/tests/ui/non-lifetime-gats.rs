//@ known-failure
use std::ops::Deref;

trait PointerFamily {
    type Pointer<T>: Deref<Target = T>;

    fn new<T>(value: T) -> Self::Pointer<T>;
}

struct BoxFamily;

impl PointerFamily for BoxFamily {
    type Pointer<T> = Box<T>;

    fn new<T>(value: T) -> Self::Pointer<T> {
        Box::new(value)
    }
}

fn make_pointer<F: PointerFamily, T>(x: T) -> F::Pointer<T> {
    F::new(x)
}

fn main() {
    let _: Box<_> = make_pointer::<BoxFamily, _>(42);
}

mod moar_variables {
    // Dummy trait to check we handle variables in clauses correctly.
    trait Link<T> {}
    impl<T, U> Link<T> for U {}

    trait Trait<T> {
        type Type<U>: Link<T>;
    }

    struct Foo;
    impl<T> Trait<Option<T>> for Foo {
        type Type<U> = (T, U);
    }
}
