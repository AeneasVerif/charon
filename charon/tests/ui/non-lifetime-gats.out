thread 'rustc' panicked at /rustc/730d5d4095a264ef5f7c0a0781eea68c15431d45/compiler/rustc_type_ir/src/binder.rs:711:9:
type parameter `T/#1` (T/#1/1) out of range when instantiating, args=[Self/#0]
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: Thread panicked when extracting item `test_crate::PointerFamily`.
 --> tests/ui/non-lifetime-gats.rs:4:1
  |
4 | trait PointerFamily {
  | ^^^^^^^^^^^^^^^^^^^

error: Ignoring the following item due to an error: test_crate::PointerFamily
 --> tests/ui/non-lifetime-gats.rs:4:1
  |
4 | trait PointerFamily {
  | ^^^^^^^^^^^^^^^^^^^

error: Could not find the type variable "T" (index: 0)
  --> tests/ui/non-lifetime-gats.rs:13:5
   |
13 |     type Pointer<T> = Box<T>;
   |     ^^^^^^^^^^^^^^^

error: Ignoring the following item due to an error: test_crate::{impl#0}
  --> tests/ui/non-lifetime-gats.rs:12:1
   |
12 | impl PointerFamily for BoxFamily {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Could not find the type variable "U" (index: 1)
  --> tests/ui/non-lifetime-gats.rs:39:9
   |
39 |         type Type<U> = (T, U);
   |         ^^^^^^^^^^^^

error: Ignoring the following item due to an error: test_crate::moar_variables::{impl#1}
  --> tests/ui/non-lifetime-gats.rs:38:5
   |
38 |     impl<T> Trait<Option<T>> for Foo {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 6 previous errors

Error: Charon driver exited with code 101
